GitHubBranches.ps1

```diff
@@ -154,6 +154,351 @@ filter Get-GitHubRepositoryBranch
+
+    .PARAMETER TargetBranchName
+        Name of the branch to be created.
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api.  Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .PARAMETER NoStatus
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+
+    .INPUTS
+        GitHub.Branch
+        GitHub.Content
+        GitHub.Event
+        GitHub.Issue
+        GitHub.IssueComment
+        GitHub.Label
+        GitHub.Milestone
+        GitHub.PullRequest
+        GitHub.Project
+        GitHub.ProjectCard
+        GitHub.ProjectColumn
+        GitHub.Release
+        GitHub.Repository
+
+    .OUTPUTS
+        GitHub.Branch
+
+    .EXAMPLE
+        New-GitHubRepositoryBranch -OwnerName microsoft -RepositoryName PowerShellForGitHub -TargetBranchName new-branch
+
+        Creates a new branch in the specified repository from the master branch.
+
+    .EXAMPLE
+        New-GitHubRepositoryBranch -Uri 'https://github.com/microsoft/PowerShellForGitHub' -BranchName develop -TargetBranchName new-branch
+
+        Creates a new branch in the specified repository from the 'develop' origin branch.
+
+    .EXAMPLE
+        $repo = Get-GithubRepository -Uri https://github.com/You/YourRepo
+        $repo | New-GitHubRepositoryBranch -TargetBranchName new-branch
+
+        You can also pipe in a repo that was returned from a previous command.
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParameterSetName = 'Elements',
+        PositionalBinding = $false
+    )]
+    [OutputType({$script:GitHubBranchTypeName})]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSShouldProcess', '',
+        Justification = 'Methods called within here make use of PSShouldProcess, and the switch is
+        passed on to them inherently.')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSReviewUnusedParameter', '',
+        Justification = 'One or more parameters (like NoStatus) are only referenced by helper
+        methods which get access to it from the stack via Get-Variable -Scope 1.')]
+    [Alias('New-GitHubBranch')]
+    param(
+        [Parameter(ParameterSetName = 'Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName = 'Elements')]
+        [string] $RepositoryName,
+
+        [Parameter(
+            Mandatory,
+            ValueFromPipelineByPropertyName,
+            Position = 1,
+            ParameterSetName = 'Uri')]
+        [Alias('RepositoryUrl')]
+        [string] $Uri,
+
+        [string] $BranchName = 'master',
```

Comment:
We should be able to pipe in a `GitHub.Branch` to this to create a new branch from it.

```suggestion
        [Parameter(ValueFromPipelineByPropertyName)]
        [string] $BranchName = 'master',
```

Reply:
Additionally, let's have a conversation similar to our TeamSlug conversation.  I feel that this should have an optional parameter for `Sha`.  If not supplied, I agree that this can continue to look up the existing Sha value.  However, in the event that a user pipes in a `GitHub.Branch` object, it should have everything that it needs.

To that end, my suggestion is that:
   1. `Add-GitHubBranchAdditionalProperties` should add `Sha` as a top-level property that it adds to a `GitHub.Branch` object.
   2. This method should have an optional `Sha` property that can receive `ValueFromPipelineByPropertyName`, and only resorts to looking up the `Sha` via an additional REST request if it wasn't supplied.

If you don't want to do this yourself, then I will do it as a follow-up check-in once this is merged in.

Reply:
Yep, as previously @HowardWolosky, can we do this on another PR.

Reply:
Tracked as #261